mpmath/functions/functions.py

```diff
@@ -238,7 +238,7 @@ def log1p(ctx, x):
             b2 = ctx.fmul(b, b, prec=2*wp)
             diff = ctx.fsub(b2, a2, prec=3*wp)
             real = a + ctx.ldexp(diff, -1)
-        result = ctx.mpc(real, b)
+        result = ctx.mpc(real, b) if b else real
```

Comment:
This will result in a real value even if the original input was complex (with imaginary part 0).  With this pull request:

```
>>> import mpmath
>>> mpmath.__version__
'1.4.0a2.dev97+g0213b04'
>>> from mpmath import mp
>>> mp.dps = 25
>>> mp.log1p(0.1 + 0j)  # Preserves complex type
mpc(real='0.09531017980432486509042041749', imag='0.0')
>>> mp.log1p(1e-30 + 0j)  # Loses complex type, result is real
mpf('1.000000000000000083336420608e-30')
```

In 1.3.0, both of the results shown above are complex.

Reply:
Probably the Kahan trick for real input could be used.  I'll remove this commit as the fix seems not trivial.